src/digest/sha1.rs

```diff
@@ -66,15 +66,48 @@ fn block_data_order_(mut H: State, M: &[[<W32 as Word>::InputBytes; 16]]) -> Sta
         let mut e = H[4];
 
         // FIPS 180-4 6.1.2 Step 3
-        for t in 0..ROUNDS {
-            // FIPS 180-4 {4.1.1, 4.2.1}
-            let (k, f) = match t {
-                0..=19 => (Wrapping(0x5a827999), ch(b, c, d)),
-                20..=39 => (Wrapping(0x6ed9eba1), parity(b, c, d)),
-                40..=59 => (Wrapping(0x8f1bbcdc), maj(b, c, d)),
-                60..=79 => (Wrapping(0xca62c1d6), parity(b, c, d)),
-                _ => unreachable!(),
-            };
+
+        // The following four loops are identical except for the values of k
+        // and f, which are as specified in FIPS 180-4 {4.1.1, 4.2.1}
+        for t in 0..20 {
```

Comment:
I like the idea but I'd still like to find a way to keep things more in 1:1 correspondence with the spec. Please try factoring out the commonality into a callable `#[inline]` or `#[inline(always)]` function that gets called four times in succession:

```
fn step3(...) {
}

let (....) = step3(0x5a827999, ch)
let (....) = step3(0x6ed9eba1, parity);
let (....) = step3(0x8f1bbcdc, maj);
let (....) = step3(0xca62c1d6, parity);
```

Reply:
Ah, yes. Shadowing the variables is a good idea. Looks much better now and a bit of extra performance fell out as well. Will clean up the patches and push.